Update Ubuntu installer

WardBenjamin opened the pull request:
Change Ruby and Rails versions, update dependencies, change Node.js to use a up-to-date PPA repository, install Heroku toolbelt from Github instead of their install script as to not install a system ruby package.

WardBenjamin committed:
Update Ubuntu installer

Change Ruby and Rails versions, update dependencies, change Node.js to use a up-to-date PPA repository, install Heroku toolbelt from Github instead of their install script as to not install a system ruby package.

WardBenjamin commented:
Please wait to merge, ironing out one last heroku bug. It installs correctly and is usable until restart, then errors with a `ruby: cannot find the heroku gem` error, unless you use the fully specified `/usr/bin/heroku` instead of just `heroku`.

The actual output if you are interested:

``````
user@BEN-PC-LINUX:~$ heroku login
/home/user/.rvm/rubies/ruby-2.2.2/lib/ruby/site_ruby/2.2.0/rubygems/dependency.rb:315:in `to_specs': Could not find 'heroku' (>= 0) among 96 total gem(s) (Gem::LoadError)
Checked in 'GEM_PATH=/home/user/.rvm/gems/ruby-2.2.2:/home/user/.rvm/gems/ruby-2.2.2@global', execute `gem env` for more information
        from /home/user/.rvm/rubies/ruby-2.2.2/lib/ruby/site_ruby/2.2.0/rubygems/dependency.rb:324:in `to_spec'
        from /home/user/.rvm/rubies/ruby-2.2.2/lib/ruby/site_ruby/2.2.0/rubygems/core_ext/kernel_gem.rb:64:in `gem'
        from /home/user/.rvm/gems/ruby-2.2.2/bin/heroku:22:in `<main>'
        from /home/user/.rvm/gems/ruby-2.2.2/bin/ruby_executable_hooks:15:in `eval'
        from /home/user/.rvm/gems/ruby-2.2.2/bin/ruby_executable_hooks:15:in `<main>'
user@BEN-PC-LINUX:~$ /usr/bin/heroku login
Enter your Heroku credentials.
Email: ```
``````

WardBenjamin commented:
I fixed this by echoing `alias heroku="/usr/bin/heroku"` to my .bashrc file. I will add that to the install script. (This can now be merged).

WardBenjamin committed:
Fix heroku command not working except in /usr/local/heroku

WardBenjamin committed:
Fix not actually installing nodejs

WardBenjamin committed:
Update PostgreSQL to use deb. repository

WardBenjamin committed:
Add heroku final install and heroku-status plugin

WardBenjamin committed:
Fix .bashrc being .bashrcro

leesharma commented on rails-install-ubuntu.sh:
```diff
@@ -2,50 +2,69 @@
 echo "Installs RVM (Ruby Version Manager) for handling Ruby installation"
 gpg --keyserver hkp://keys.gnupg.net --recv-keys 409B6B1796C275462A1703113804BB82D39DC0E3
-curl -sSL https://get.rvm.io | bash -s -- --version 1.26.11
+curl -sSL https://get.rvm.io | bash -s stable
```
While the stable version of RVM works at the moment, this will break with future releases. RVM signs and verifies all releases, and I believe the key is specific to the release (see [rvm security](https://rvm.io/rvm/security)).

It's not great for automation, so if you can come up with a better way, let me know. Otherwise, this should be changed back.

WardBenjamin replied:
Good catch, I agree that this should be changed back. I don't really know how this could be automated save a website scrape from the rvm website but that is way out of scope for this script.

leesharma commented on rails-install-ubuntu.sh:
```diff
@@ -2,50 +2,69 @@
 echo -n "Should be sqlite 3.8.2 or higher: sqlite "
 sqlite3 --version
-echo -n "Should be rvm 1.26.11:          "
+echo -n "Should be rvm 1.26.11 or higher:         "
```
See above comment about rvm.

leesharma commented on rails-install-ubuntu.sh:
```diff
@@ -2,50 +2,69 @@
 set -e
 
+echo "Adding PPA for up-to-date Node.js runtime. Give your password when asked."
+sudo add-apt-repository ppa:chris-lea/node.js
```
Good call on Node; I forgot about the js runtime. 

Why are you using a third-party repo (chris-lea ppa) instead of:

```
curl -sL https://deb.nodesource.com/setup | sudo bash -
sudo apt-get install -y nodejs
```

?

WardBenjamin replied:
Huh, I forgot there was an updated version. If you take a look at that script you linked, it actually removes this PPA if present, because this was the official one before they made their own `deb` repository. I'll flip it around to that.

leesharma commented on rails-install-ubuntu.sh:
```diff
@@ -2,50 +2,69 @@
+echo "Adding PPA for up-to-date Node.js runtime. Give your password when asked."
+sudo add-apt-repository ppa:chris-lea/node.js
+
+echo "Adding deb. repository for PostgreSQL"
```
I'm not running Ubuntu myself, and I didn't find an explanation in the commit messages. Did the Postgres setup below not work? What is the reason for this change?

WardBenjamin replied:
The Postgres setup below worked, however as with the Node.js repository, the default Ubuntu packages are outdated by usually several months, if not more, since they have to go through a long application and testing process. The deb. repositories and PPAs solve this by being able to download packages right from the people who actually make them.

leesharma commented on rails-install-ubuntu.sh:
```diff
@@ -2,50 +2,69 @@
 
 echo "Installs packages. Give your password when asked."
-sudo apt-get install build-essential bison openssl libreadline6 libreadline6-dev curl git-core zlib1g zlib1g-dev libssl-dev libyaml-dev libsqlite3-0 libsqlite3-dev sqlite3 postgresql postgresql-contrib libpq-dev pgadmin3 libxml2-dev libxslt-dev autoconf libc6-dev nodejs -y
+sudo apt-get install build-essential git-core curl openssl libssl-dev libcurl4-openssl-dev zlib1g zlib1g-dev libreadline libreadline-dev libreadline6 libreadline6-dev libyaml-dev libsqlite3-dev libsqlite3-0 sqlite3 libxml2-dev libxslt1-dev python-software-properties libffi-dev libgdm-dev libncurses5-dev automake autoconf libtool bison libc6-dev postgresql-common postgresql-9.4 libpq-dev pgadmin3  ...
```
There were several package additions here. Can you explain the reasons for each change?

WardBenjamin replied:
The reasons are as follows:
Added libreadline because of oddities when only the new version (libreadline6) is installed (this if for other programs, so it could be removed).
Added automake, ncurses, and a few others for gem building/installation (again, could be removed if needed)
Updated all of the Postgres packages to use the ones from the deb. repository.

I don't know how python-software-properties made it's way in there. It's used mainly for add-apt-repository but I don't think anything in this script is directly using it. I'll remove that.

WardBenjamin committed:
Revert RVM version change, add Node.js deb repo

Also removed extra double quote character from the PostgreSQL installation.

WardBenjamin committed:
Remove python-software-properties and others

I have removed psp and a few other packages that are used to build gems but are not necessarily needed for the initial Rails setup.
